web/src/copy_and_paste.js

```diff
@@ -294,29 +294,165 @@ function get_end_tr_from_endc($endc) {
+        // to extract and add the language of the code block (if any) to the fence.
+        replacement(content, node, options) {
+            const language = node.closest(".codehilite")?.dataset?.codeLanguage || "";
+
+            const fenceChar = options.fence.charAt(0);
+            let fenceSize = 3;
+            const fenceInCodeRegex = new RegExp("^" + fenceChar + "{3,}", "gm");
+
+            let match;
+            while ((match = fenceInCodeRegex.exec(content))) {
+                if (match[0].length >= fenceSize) {
+                    fenceSize = match[0].length + 1;
+                }
+            }
+
+            const fence = fenceChar.repeat(fenceSize);
+
+            return (
+                "\n\n" +
+                fence +
+                language +
+                "\n" +
+                content.replace(/\n$/, "") +
+                "\n" +
+                fence +
+                "\n\n"
+            );
+        },
+    });
+    turndownService.addRule("zulipImagePreview", {
         filter(node) {
+            // select image previews in Zulip messages
             return (
-                node.nodeName === "A" && node.href === node.innerHTML && node.href === node.title
+                node.classList.contains("message_inline_image") && node.firstChild.nodeName === "A"
             );
         },
-        replacement(content) {
-            return content;
+
+        replacement(content, node) {
+            // We parse the copied html to then check if the generating link (which, if
+            // present, always comes before the preview in the copied html) is also there.
+
+            // If the preview has an aria-label, it means it does have a named link in the
+            // message, and if the 1st element with the same image link in the copied html
+            // does not have the `message_inline_image` class, it means it is the generating
+            // link, and not the preview, meaning the generating link is copied as well.
+            const copied_html = new DOMParser().parseFromString(paste_html, "text/html");
+            if (
+                node.firstChild.ariaLabel &&
+                !copied_html
+                    .querySelector("a[href='" + node.firstChild.getAttribute("href") + "']")
+                    ?.parentNode?.classList.contains("message_inline_image")
+            ) {
+                // We skip previews which have their generating link copied too, to avoid
+                // double pasting the same link.
+                return "";
+            }
+            return image_to_zulip_markdown(content, node.firstChild);
+        },
+    });
+    turndownService.addRule("images", {
+        filter: "img",
+
+        replacement: image_to_zulip_markdown,
+    });
+    turndownService.addRule("math", {
+        // We don't have a way to get the original LaTeX code from the rendered
+        // `math` so we drop it to avoid pasting gibberish.
```

Comment:
I think this comment might be wrong -- in https://github.com/zulip/zulip/issues/25289, I'm pretty sure we figured out a way to get the LaTeX source out of KaTeX HTML.

Reply:
The method in #25289 isn't quite going to work, because in the HTML it's reaching into, syntax like `\int_{0}^{1}` will be presented as the characters `∫01`, which Katex is going to present literally.

But.

I found that there's an `<annotation>` inside the actual `<math>` element that contains the original string as would be passed to Markdown, and if we use CSS to display it as a block, 1) it doesn't seem to actually show on screen (though that needs a LOT of confirmation), but 2) it becomes part of the HTML payload that we can then crawl through the DOM to open up and extract in custom copy-paste logic.

Reply:
The `annotation` html element unfortunately isn't part of the html that is copied, so there's no trace of the raw LaTeX in the copied html

Reply:
That's because it's hidden by default, I believe. If you experiment with setting it to `display: block`, you'll find it gets picked up in the copy function. What's surprising and would need to be reasoned through, if we go this route, is to make sure it doesn't render to the screen. 

There's also the question about what happens if a user goes to paste Zulip blocks like this into another application, of course.

Reply:
You're right! Making `annotation`'s display block, lets it get copied. And it still doesn't render on screen (because of katex library's CSS accessibility hack to only show its ancestor element `katex-mathml` to screen readers)

Reply:
Ah, cool. So I think the other thing I found was you need to then figure out whether it's an inline or block math element you're dealing with, but it's a quick trip through the DOM to do that.

Reply:
Right, here's what I have currently:
```
    turndownService.addRule("zulipLaTeX", {
        filter(node) {
            // select LaTeX in Zulip messages
            return node.nodeName === "SPAN" && node.classList.contains("katex") &&
            node.firstChild.nodeName === "SPAN" && node.firstChild.classList.contains("katex-mathml") &&
            node.firstChild.querySelector("annotation") !== null;
        },

        replacement(content, node) {
            const raw_latex = node.querySelector("annotation").innerHTML;
            console.log(raw_latex)
            if (node.querySelector("math").getAttribute("display") === "block") {
                return "\`\`\`math\n" + raw_latex + "\n\`\`\`";
            }
            return "$$" + raw_latex + "$$";
        },
    });
```

But we'll probably need to add a zulip specific class to make sure this only works for zulip LaTeX, and maybe retain the other rule (for just not showing math elements) for LaTeX from outside :thinking: